lib/events/index.js

```diff
@@ -1958,56 +1987,166 @@ export default function (self) {
+          rowIndex = 0;
+        }
+
+        const overfillingRow = rowPosReal >= rows.length;
+        const rowPosition = reverseRows
+          ? rowsLength - rowPosReal - 1
+          : rowPosReal;
         // Rows may have been moved by user, so get the actual row index
         // (instead of the row index at which the row is rendered):
-        var realRowIndex = self.orders.rows[startRowIndex + rowIndex];
+        var realRowIndex = self.orders.rows[startRowIndex + rowPosition];
         var cells = rows[rowIndex];
+        const cellsLength = Math.max(cells.length, minColumnsLength);
 
         var existingRowData = self.viewData[realRowIndex];
         var newRowData = Object.assign({}, existingRowData);
 
         selections[realRowIndex] = [];
         self.selectedRows.push(realRowIndex);
 
-        for (var colIndex = 0; colIndex < cells.length; colIndex++) {
-          var column = schema[startColIndex + colIndex];
+        const fillDownArgs = cellCallback
+          ? {
+              source: source,
+              rowData: newRowData,
+              existingRowData: existingRowData,
+              rowIndex: realRowIndex,
+              rows: rows,
+              rowPosition: rowPosReal,
+              rowsLength: rowsLength,
+              cells: cells,
+              direction: direction,
+              reversed:
+                direction === 'horizontal' ? reverseColumns : reverseRows,
+              overfillingRow: overfillingRow || isHandle,
+              overfillingRowPosition: isHandle
+                ? rowPosReal
+                : overfillingRow
+                ? rowPosReal - rows.length
+                : -1,
+            }
+          : undefined;
+
+        for (
+          var colPosReal = 0, colIndex = 0;
+          colPosReal < cellsLength;
+          colPosReal++, colIndex++
+        ) {
+          if (colIndex >= cells.length) {
+            colIndex = 0;
+          }
+
+          const overfillingColumn = colPosReal >= cells.length;
+          const colPosition = reverseColumns
+            ? cellsLength - colPosReal - 1
+            : colPosReal;
+          var column = schema[startColIndex + colPosition];
 
           if (!column) {
             console.warn('Paste data exceeded grid bounds. Skipping.');
             continue;
           }
 
           var columnName = column.name;
-          var cellData = cells[colIndex].value
-            .map((item) => item.value)
-            .join('');
-
-          if (cellData === undefined || cellData === null) {
-            newRowData[columnName] = existingRowData[columnName];
-            continue;
+          var cellData = isPasting
+            ? cells[colIndex].value.map((item) => item.value).join('')
+            : cells[colIndex];
+          const existingCellData = existingRowData[columnName];
+
+          if (cellCallback) {
+            newRowData[columnName] = cellCallback(
+              Object.assign(fillDownArgs, {
```

Comment:
I'm pretty sure we can use the spread operator here to make our life a bit easier, as well as the object property shorthand (not having to do `{ foo: foo }` but just `{ foo }`:

```js
newRowData[columnName] = cellCallback({
  ...fillDownArgs,
  column,
  columnIndex: startColIndex + colPosition,
  existingCellData,
  // etc
});
```

Reply:
Got it. Thanks!